vllm/engine/llm_engine.py

```diff
@@ -562,36 +607,368 @@ def _add_processed_request(
         prompt_adapter_request: Optional[PromptAdapterRequest] = None,
     ) -> LLMInputs:
+        '''
+        For decoder-only models:
+        Process an input prompt
+        into an `LLMInputs` instance.
+
+        Arguments:
+
+        * inputs: input prompt
+        * lora_request
+        * request_id
+        * prompt_adapter_request
+
+        Returns:
+
+        * `LLMInputs` instance
+        '''
+
         if isinstance(inputs, str):
             inputs = {"prompt": inputs}
+        prompt = inputs.get("prompt")
 
         if "prompt_token_ids" not in inputs:
-            tokenizer = self.get_tokenizer_group("prompts must be None if "
-                                                 "skip_tokenizer_init is True")
-
-            prompt_token_ids = tokenizer.encode(request_id=request_id,
-                                                prompt=inputs["prompt"],
-                                                lora_request=lora_request)
+            prompt_token_ids = self._tokenize_prompt(
+                prompt,
+                request_id=request_id,
+                lora_request=lora_request,
+            )
         else:
             prompt_token_ids = inputs["prompt_token_ids"]
 
         if prompt_adapter_request:
-            prompt_token_ids = \
-                [0] * prompt_adapter_request.prompt_adapter_num_virtual_tokens\
-                         + prompt_token_ids
+            prompt_token_ids = (
+                [0] * prompt_adapter_request.prompt_adapter_num_virtual_tokens
+                + prompt_token_ids)
 
-        llm_inputs = LLMInputs(prompt_token_ids=prompt_token_ids,
-                               prompt=inputs.get("prompt"),
-                               multi_modal_data=inputs.get("multi_modal_data"))
+        return LLMInputs(prompt_token_ids=prompt_token_ids,
+                         prompt=prompt,
+                         multi_modal_data=inputs.get("multi_modal_data"))
 
-        return self.input_processor(llm_inputs)
+    def process_model_inputs(
+        self,
+        request_id: str,
+        inputs: PromptInputs,
+        lora_request: Optional[LoRARequest] = None,
+        prompt_adapter_request: Optional[PromptAdapterRequest] = None,
+    ) -> LLMInputs:
+
+        if self.is_encoder_decoder_model():
+            # Encoder-decoder model requires special mapping of
+            # input prompts to encoder & decoder
+            return self.input_processor(
+                self._process_encoder_decoder_prompt(
+                    inputs,
+                    request_id=request_id,
+                ))
+
+        else:
+            # Decoder-only operation
+            return self.input_processor(
+                self._process_decoder_only_prompt(
+                    inputs,
+                    request_id=request_id,
+                    lora_request=lora_request,
+                    prompt_adapter_request=prompt_adapter_request,
+                ))
```

Comment:
Maybe get the inputs here and have a single call to
```python
return self.input_processor(model_inputs)
```
at the end?

Reply:
Agreed. Now the method body looks like

```
        if self.is_encoder_decoder_model():
            # Encoder-decoder model requires special mapping of
            # input prompts to encoder & decoder
            
            model_inputs = self._process_encoder_decoder_prompt(
                inputs,
                request_id=request_id,
            )
        else:
            # Decoder-only operation
            model_inputs = self._process_decoder_only_prompt(
                    inputs,
                    request_id=request_id,
                    lora_request=lora_request,
                    prompt_adapter_request=prompt_adapter_request,
                )
    
        return self.input_processor(model_inputs)
```

Reply:
I considered assigning model_inputs in a one-liner but I think that would be overwhelming to someone reading the code. This seems pretty readable.